Template tensor products

dschwen opened the pull request:
Closes #21226

This keeps the old methods around for backwards compatibility (they wrap the new methods).

dschwen committed:
Template products (#21226)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1069487/) on 4cd378c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d444c549368b641223dca5ffd90e07d934b1f99a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1069562/) on 53663a6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d444c549368b641223dca5ffd90e07d934b1f99a`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1069614/) on 971b701 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21227/site)

This comment will be updated on new commits.

dschwen commented on framework/include/utils/RankFourTensor.h:
```diff
@@ -757,3 +779,20 @@ RankFourTensorTempl<T>::inverse() const
+    for (x[1] = 0; x[1] < N; ++x[1])
+      for (x[2] = 0; x[2] < N; ++x[2])
+        for (x[3] = 0; x[3] < N; ++x[3])
+          result(z[y][0], z[y][1], z[y][2]) += (*this)(x[0], x[1], x[2], x[3]) * b(x[y]);
```
Ugh. Should be x[...] For every index for result

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1070177/) on 94cc356 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 65228e22d11d97b0b355178c64ae457c350f8134`

dschwen committed:
Some use (#21226)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1070310/) on 9a44594 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>65228e</th>
      <th colspan=4>#21227 9a4459</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.57%</td>
      <td>82.60%</td>
      <td>+0.03%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>75542</td>
      <td>75538</td>
      <td>-4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15947</td>
      <td>15915</td>
      <td>-32</td>
      <td>43</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21227/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21227/coverage/framework)


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>65228e</th>
      <th colspan=4>#21227 9a4459</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.54%</td>
      <td>89.54%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26042</td>
      <td>26042</td>
      <td>-</td>
      <td>19</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3042</td>
      <td>3042</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21227/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21227/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21227/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

hugary1995 commented:
Cool!

dschwen committed:
Lower case indices (#21226)

dschwen committed:
Convert more calls, remove old methods (#21226)

hugary1995 commented:
This is debatable: When I think of "mixed product", I always think of it as the mixed product of a and b. Hence it would be natural to use a signature like `MathUtils::mixedProduct<i,j,k,l>(a, b)`. But that didn't happen for obvious reasons... Now while you are at it, I think it is more natural to call it something like `times` or `otimes` -- they are shorter names, and they properly represent the operator name between a and b, e.g. `a.times<i,j,k,l>(b)` looks and reads better (to me) than `a.mixedProduct<i,j,k,l>(b)`. Plus, if you search online, there is no such a thing called "mixed product"...

dschwen commented:
Sure, I can rename the templates method to `times<>`.
Another question would be whether to more clearly separate the two sets of Indices, e.g. by wrapping them in a separate type (`times<Indices<i,j>, Indices<k,l>>`).

dschwen commented on modules/tensor_mechanics/unit/src/RankTwoTensorTest.C:
```diff
@@ -540,9 +540,10 @@ TEST_F(RankTwoTensorTest, ADConversion)
 }
 
-TEST_F(RankTwoTensorTest, mixedProductIjJklm)
+TEST_F(RankTwoTensorTest, mixedProductIMMJKL)
```
This test passes on new compilers but gives wrong results on min clang.On every other compiler (including gcc min) the template instantiation `mixedProduct<i, m, m, j, k, l>` works fine. All other tested instantiations of this same template work on all compilers.

What the heck?

dschwen committed:
Work around min clang scope resolution bug (#21226)

clang 5.0.2 uses the value for `m` from the enum, rather than from the
template parameter. Working around it by not having a template parameter
named m.

dschwen committed:
Rename methods (#21226)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1072929/) on fbca3a6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f6060e2047d3c8a125e251bb6dc4a773cc34176b`

dschwen committed:
Use trailing underscore to avoid identifier collision (#21226)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1074306/) on 559615b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b1d9e7e11abdf9117ab26c5e04ba030c4eed5ff7`

dschwen committed:
Fix index mixup and use indices locally (#21226)

dschwen committed:
Remove old signature (#21226)

dschwen commented:
@hugary1995 do you have any more comments?

hugary1995 commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -315,29 +316,23 @@ class RankTwoTensorTempl : public TensorValue<T>
-
-  /// returns C_ijkl = a_il * b_jk
-  RankFourTensorTempl<T> mixedProductIlJk(const RankTwoTensorTempl<T> & a) const;
+  /// returns C_ijkl = a_no * b_pq
+  template <int n, int o, int p, int q>
+  RankFourTensorTempl<T> times(const RankTwoTensorTempl<T> & b) const;
```
The doxygen comment might need to be changed. While I know what you mean here, others reading this the first time may be confused, because it seems like we are creating indices i, j, k, l out of indices n, o, p, q, which is...magic.

This might be more accurate:
```
Assuming a bijective index map M: {n,o,p,q} -> {i,j,k,l}, this function returns C_{ijkl} = a_{M(n)M(o)} b_{M(p)M(q)}.
```

dschwen replied:
This is great

hugary1995 commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -315,29 +316,23 @@ class RankTwoTensorTempl : public TensorValue<T>
 
-  /// returns C_iklm = a_ij * b_jklm
-  RankFourTensorTempl<T> mixedProductIjJklm(const RankFourTensorTempl<T> & a) const;
+  /// returns C_ijkl = a_no * b_pqrs
+  template <int n, int o, int p, int q, int r, int s>
+  RankFourTensorTempl<T> times(const RankFourTensorTempl<T> & b) const;
```
Same comment as above.

dschwen replied:
surjective in this case. added.

hugary1995 reviewed (commented):
I'm being real picky about doxygen comments. Everything else looks great.

dschwen committed:
Address comments (#21226)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1075520/) on 036f71b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21227/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0973fa34bb182ce933a1991d1fd02cafaa169d80`

dschwen commented:
> Your code requires style changes.

I suppose I haven't set up my VSCode quite right yet...

bwspenc commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -315,29 +316,31 @@ class RankTwoTensorTempl : public TensorValue<T>
+  /**
+   * Assuming a bijective index map M: {n,o,p,q} -> {i,j,k,l}, this function returns
+   * C_{ijkl} = a_{M(n)M(o)} b_{M(p)M(q)}. The bijective map is deduced from the
+   * positions of template arguments n, o, p, q.
```
I have to admit that I had no idea what "bijective" and "surjective" maps are. Is it possible to elaborate on these comments a little? I'm concerned that it's not going to be very obvious to people how to use these.

hugary1995 replied:
I think this sort of method is better demonstrated with example usages. While we can include a simple example in the doxygen comment, I think long-term we would want to have a .md page dedicated for such. I have created an issue #21330, and I will find a chance to work on it in the next few days.

hugary1995 replied:
Regarding the name, I think I'm the one who made Daniel change it to `times` from `mixedProduct`... You can see my comments above, but `tensorProduct` is certainly better than `mixedProduct` (at least it's not a random operator name that we came up with).

dschwen replied:
Where mnop are a permutation of Ijkl...

dschwen replied:
I don't think `tensorProduct` is any better.  The participating types imply this already. But it's not worse either, so I'll make the change.

dschwen replied:
>  I'm mostly thinking that it would be easier to search for to find examples than times.

Sheesh. We're all using VSCode now and don't have to resort to lowly "find in files" we can use intellisense to show all uses of that function...

dschwen replied:
So should I rename `timesTranspose` to `tensorProductWithTransposeTensor`?

hugary1995 replied:
You should write a proposal to C++23 to add support for new operators `operator*T` and `operatorT*`.

hugary1995 replied:
But seriously though, the named constructors `timesTranspose`, `transposeTimes` and `plusTranspose` seem a little bit unnecessary to me. Even in the case of symmetric tensors, I don't think they buy us much efficiency.

hugary1995 replied:
Oh I think I get it now -- you need these methods to guarantee symmetry. Then yeah, I've no idea how to rename it... `tensorProductWithTransposeTensor` seems to long.

bwspenc replied:
Sounds like `times` is as good as anything -- let's just go with that.

reverendbedford commented:
I guess enough people have weighed in on this, but:

1. Why do we have separate `contraction` and `times` methods?
2. Why can't we just call both `contraction` and `times` `einsum` like most other libraries do?
3. But if people don't like #2 I agree that `tensorProduct` is is better than `times`.

dschwen commented:
@reverendbedford because this is not a generic Einstein summation capability! Those do exist, but IMO their C++ syntax is very offputting. In particular this PR has no logic to check for duplicate indices and missing indices.

dschwen committed:
ELI5 (#21226)

reverendbedford commented:
> In particular this PR has no logic to check for duplicate indices and missing indices.

Ah, fair enough then.  I thought by specifying, as you are, the rank of the inputs + the number of indices involved in the contraction that you tacitly providing both of these but maybe not.  I didn't think about it very hard.
